src/storage/src/compaction/scheduler.rs

```diff
@@ -0,0 +1,455 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::marker::PhantomData;
+use std::sync::{Arc, Mutex};
+
+use async_trait::async_trait;
+use common_telemetry::{debug, info};
+use snafu::ResultExt;
+use table::metadata::TableId;
+use tokio::sync::{Notify, RwLock};
+use tokio::task::JoinHandle;
+use tokio_util::sync::CancellationToken;
+
+use crate::compaction::dedup_deque::DedupDeque;
+use crate::compaction::picker::Picker;
+use crate::compaction::rate_limit::{
+    BoxedRateLimitToken, CascadeRateLimiter, MaxInflightTaskLimiter, RateLimitToken, RateLimiter,
+};
+use crate::compaction::task::CompactionTask;
+use crate::error::{Result, StopCompactionSchedulerSnafu};
+
+/// Table compaction request.
+#[derive(Default)]
+pub struct CompactionRequestImpl {
+    table_id: TableId,
+}
+
+impl CompactionRequest for CompactionRequestImpl {
+    #[inline]
+    fn table_id(&self) -> TableId {
+        self.table_id
+    }
+}
+
+pub trait CompactionRequest: Send + Sync + Default + 'static {
+    fn table_id(&self) -> TableId;
+}
+
+#[derive(Debug)]
+pub struct CompactionSchedulerConfig {
+    max_inflight_task: usize,
+}
+
+impl Default for CompactionSchedulerConfig {
+    fn default() -> Self {
+        Self {
+            max_inflight_task: 16,
+        }
+    }
+}
+
+/// CompactionScheduler defines a set of API to schedule compaction tasks.
+#[async_trait]
+pub trait CompactionScheduler<R> {
+    /// Schedules a compaction request.
+    /// Returns true if request is scheduled. Returns false if task queue already
+    /// contains the request with same table id.
+    async fn schedule(&self, request: R) -> Result<bool>;
+
+    /// Stops compaction scheduler.
+    async fn stop(&self) -> Result<()>;
+}
+
+/// Compaction task scheduler based on local state.
+#[allow(unused)]
+pub struct LocalCompactionScheduler<R: CompactionRequest> {
+    request_queue: Arc<RwLock<DedupDeque<TableId, R>>>,
```

Comment:
Perhaps std rwlock or parking_lot::RwLock could be used to provide better performance

Reply:
Good point. Is it necessary to use an async rwlock?

Reply:
This lock won't be held between tasks, so it's safe to replace it with sync lock